java/core/src/main/java/org/eclipse/tahu/mqtt/TahuClient.java

```diff
@@ -0,0 +1,1276 @@
+		this.onlineDate = new Date();
+	}
+
+	/**
+	 * Returns a {@link Date} instance representing the offline date.
+	 * 
+	 * @return the offline date.
+	 */
+	public Date getOfflineDateTime() {
+		return this.offlineDate;
+	}
+
+	/**
+	 * Renews the offline date.
+	 */
+	public void renewOfflineDate() {
+		this.offlineDate = new Date();
+	}
+
+	public IMqttDeliveryToken publish(String topic, byte[] payload, int qos, boolean retained) throws TahuException {
+		try {
+			if (client == null) {
+				throw new TahuException(TahuErrorCode.INTERNAL_ERROR,
+						"MQTT client: " + clientId.getMqttClientId() + " is null");
+			} else if (client.isConnected()) {
+				logger.debug("{}: Publishing on Topic {}, Payload Size = {}", getClientId(), topic, payload.length);
+				return client.publish(topic, payload, qos, retained);
+			} else {
+				throw new TahuException(TahuErrorCode.INTERNAL_ERROR,
+						"MQTT client: " + clientId.getMqttClientId() + " is not connected");
+			}
+		} catch (Exception e) {
+			throw new TahuException(TahuErrorCode.INTERNAL_ERROR, e);
+		}
+	}
+
+	public void asyncPublish(String topic, byte[] payload, int qos, boolean retained) throws TahuException {
+		Thread t = new Thread(new AsyncPublisher(topic, payload, qos, retained, false, 0, 0));
+		t.start();
+	}
+
+	public void asyncPublish(String topic, byte[] payload, int qos, boolean retained, boolean retry, long retryDelay,
+			int numAttempts) throws TahuException {
+		Thread t = new Thread(new AsyncPublisher(topic, payload, qos, retained, retry, retryDelay, numAttempts));
+		t.start();
+	}
+
+	/**
+	 * Subscribes to a topic.
+	 * 
+	 * @param topic the topic.
+	 * @param qos the quality of service (0, 1, or 2)
+	 * 
+	 * @return the granted QoS for the subscription
+	 * @throws TahuException
+	 */
+	public int subscribe(String topic, int qos) throws TahuException {
+		synchronized (clientLock) {
+			if (client != null) {
+				if (client.isConnected()) {
+					try {
+						logger.debug("{}: on connection to {} - Attempting to subscribe on topic {} with QoS={}",
+								getClientId(), getMqttServerName(), topic, qos);
+						IMqttToken token = client.subscribe(topic, qos);
+						logger.trace("Waiting for subscription on {}", topic);
+						token.waitForCompletion();
+						logger.trace("Done waiting for subscription on {}", topic);
+						subscriptions.put(topic, qos);
+						int[] grantedQos = token.getGrantedQos();
+						logger.debug("Granted QoS for subcription on {}: {}", topic, grantedQos[0]);
+						if (grantedQos != null && grantedQos.length == 1) {
+							return grantedQos[0];
+						} else {
+							String errorMessage = getClientId() + ": on connection to " + getMqttServerName()
+									+ " - Failed to subscribe to " + topic;
+							logger.error(errorMessage);
+							throw new TahuException(TahuErrorCode.NOT_AUTHORIZED, errorMessage);
+						}
+					} catch (MqttException e) {
+						logger.error(getClientId() + ": on connection to " + getMqttServerName()
```

Comment:
Do we care about log formatting at this point? According to the link you sent (https://stackoverflow.com/questions/8726122/java-string-concatenation-efficiency), string concatenation is efficient as well.

Reply:
I converted all of these to use concatenation so I didn't have to pull in StringUtils.